In [ ]:
import pandas as pd
import numpy as np
%matplotlib inline
import otter
grader = otter.Notebook()

# A1 | Week 3: Practice While Loops

------

### Instructions:

- Complete 50 points worth of questions to pass the assessment
  - You can attempt any number of questions and in any order provided you pass at least 50 points.
- Develop and check your answers in the spaces provided.
  - Do not remove any variables other provided markings already present in the answer spaces.
  - Do not make any changes to this notebook outside of the answer spaces provided.
  
#### Submitting

- Upload your .ipynb file to Gradescope (Make sure to restart the kernel and run all before you submit)
- You can submit as many times as needed
- When reviewing results, ignore any results listed under "Public Tests"

For more information, see the assessment page. We have also provided an example question below for clarity. 

#### Example Question
Write a Python program to display the first 10 integer numbers starting with 1 using while loops. Create the variable `example_output` and use it inside the while loop to construct the string that you want to display at the end of the program.

```
#This solution is for demonstration purposes ONLY. 
#There may be other ways of doing this task using while loops that produce the same output.
#You have the freedom to use a solution that you prefer as long as it uses while loop and 
#produces the expected output using the example_output variable.
example_output = ""
example_count = 1
while example_count <= 10:
    example_output += str(example_count)
    example_output += "\n"
    example_count += 1
example_output = example_output.strip() #this is to remove the extra new line in 10
print(example_output)
```

#### Question 1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(1 Point)
Write a Python program to display integer numbers starting with 190 to 200 using while loops. Create the variable `output1` and use it inside the while loop to construct the string that you want to display at the end of the program.

Expected Output:   
190  
191  
192  
193  
194  
195  
196  
197  
198  
199  
200

In [24]:
output1 = ""

# Write your solution here
a = 190

while a <= 200:
    output1 = output1 + str(a) + "\n"
    a += 1

output1 = output1.strip()
print(output1)

190
191
192
193
194
195
196
197
198
199
200


#### Question 2&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(1 Point)
Write a Python program to display the first 10 even numbers starting with 0 using while loops. Create the variable `output2` and use  it inside the while loop to construct the string that you want to display at the end of the program.

*Note: The displayed string should be in the same format as in Q1. For clarity, the expected output is shown below.*

Expected Output:
<p>
0<br> 
2<br> 
4<br> 
6<br> 
8<br> 
10<br> 
12<br>
14<br> 
16<br> 
18
</p>

In [1]:
output2 = ""

# Write your solution here
a = 0
i = 0
while a < 10:
    if (i % 2 == 0):
        output2 = output2 + str(i) + "\n"
  
        a += 1
        i += 1
    else:
        i += 1
output2 = output2.strip()
print (output2)

0
2
4
6
8
10
12
14
16
18


#### Question 3&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(1 Point)
Write a Python program to display the following series using while loops: `30, 40, 50, 60, 70, 80, 90, ..., 300`. Create the variable `output3` and use  it inside the while loop to construct the string that you want to display at the end of the program.

*Note: The displayed string should be in the same format as in Q1. For clarity, the expected output is shown below.*

Expected Output:
<p>
30<br> 
40<br> 
50<br> 
60<br> 
70<br> 
80<br> 
90<br> 
...<br> 
300
</p>

In [25]:
output3 = ""

# Write your solution here

a = 30

while a <= 300:
    output3 = output3 + str(a) + "\n"
    a += 10

output3 = output3.strip()
print (output3)


30
40
50
60
70
80
90
100
110
120
130
140
150
160
170
180
190
200
210
220
230
240
250
260
270
280
290
300


#### Question 4&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(1 Point)
Write a Python program to display the squares of integer numbers divisible by 4 until 200 (inclusive) using while loops. Create the variable `output4` and use it inside the while loop to construct the string that you want to display at the end of the program. (Expected output series: 16, ..., 40000)

*Note: The displayed string should be in the same format as in Q1.*

In [28]:
output4 = ""

# Write your solution here

a = 1

while a <= 200:
    if (a % 4 == 0):
        output4 = output4 + str(a**2) + "\n"
        a += 1
    else:
        a += 1
        
output4 = output4.strip()
print (output4)


16
64
144
256
400
576
784
1024
1296
1600
1936
2304
2704
3136
3600
4096
4624
5184
5776
6400
7056
7744
8464
9216
10000
10816
11664
12544
13456
14400
15376
16384
17424
18496
19600
20736
21904
23104
24336
25600
26896
28224
29584
30976
32400
33856
35344
36864
38416
40000


#### Question 5&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(1 Point)
Write a Python program to print the following series using while loops: `120, 116, 112, ..., 36`. Create the variable `output5` and use it inside the while loop to construct the string that you want to display at the end of the program.

*Note: The displayed string should be in the same format as in Q1.*

In [29]:
output5 = ""

# Write your solution here

i = 120

while i >= 36:
    output5 = output5 + str(i) + "\n"
    i -= 4

output5 = output5.strip()
print (output5)


120
116
112
108
104
100
96
92
88
84
80
76
72
68
64
60
56
52
48
44
40
36


#### Question 6&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(3 Points)
Write a Python program that adds all the odd numbers until 300 (inclusive) using while loops. Create the variable `ouput6` and use it inside the while loop for the calculations needed. Print `output6` at the end of the program to check if it works correctly.

In [30]:
output6 = 0

# Write your solution here

i = 0

while i <= 300:
    if (i % 2 == 1):
        output6 = output6 + i
        i += 1
    else:
        i += 1
print (output6)


22500


#### Question 7&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(3 Points) 
Write a Python program to calculate the total of first 10 integer numbers using while loops. However, if the number is divisible by 5, do not use them to calculate the total value. Create the variable `output7` and use it inside the while loop for the calculations needed. [Expected output: 40]
***
<details>
  <summary><span style="color:blue">Click to see Hints</span></summary>
    
* You will need to use `continue` statement. This statement rejects all the remaining statements in the current iteration of the loop and moves the control back to the top of the loop.
</details>

In [31]:
output7 = 0

# Write your solution here
i = 0

while i < 10:
    if i % 5 == 0:
        output7 = output7
        i += 1
    else:
        output7 = output7 + i
        i += 1
print (output7)
        


40


#### Question 8&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(3 Points) 
Write a Python program to calculate the total of first 20 integer numbers using while loops. However, if the program encounters a number that is divisible by 10 between 1 to 20, stop looping at the very moment and  calculate the total value until that point (i.e., excluding the number that is divisible by 10 and the numbers after that are also excluded). Create the variables `output8` and use it in the program for the calculations needed. [Expected output: 45]
***
<details>
  <summary><span style="color:blue">Click to see Hints</span></summary>
    
* You will need to use `break` statement. This will terminate the current loop.
</details>

In [32]:
output8 = 0

# Write your solution here

i = 1

while i <= 20:
    if (i % 10 == 0):
        output8 = output8
        break
    else:
        output8 = output8 + i
        i += 1

print (output8)


45


#### Question 9&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(3 Points)
Write a Python program to display the average (rounded to two decimal places) of all the numbers which are divisible by 13 but not by 3 between 300 and 600.(exclusive both numbers) using while loop. Use the variable `output9` to assign the solution to it.

In [1]:
# Write your solution here

i = 301
sum = 0
count = 0

while i > 300 and i < 600:
    if i % 13 == 0 and i % 3 == 1:
        sum = sum + i
        count += 1
        i += 1
    else:
        i += 1

output9 = sum/count
output9 = round (output9, 2)
print(output9)
    
    



461.5


#### Question 10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(3 Points)
Write a Python program to display the following pattern using while loops. Create the variable `output10` and use it inside the while loop to construct the string that you want to display at the end of the program.
<p>
*<br>
**<br>
***<br>
****<br>
*****<br>
</p>

In [47]:
output10 = ""

# Write your solution here
#4 level stars

i = 1
j = 1

while i <= 5:
    j = 1
    while j <= i:
        output10 = output10 + "*"
        j += 1
    output10 = output10 + "\n"
    i += 1
output10 = output10.strip()
print(output10)

*
**
***
****
*****


#### Question 11&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points)
Write a Python program to display the following pattern using while loops. Create the variable `output11` and use it inside the while loop to construct the string that you want to display at the end of the program.
<p>
1<br>
22<br>
333<br>
4444<br>
55555<br>
666666<br>
7777777<br>
88888888<br>
999999999<br>
</p>

In [56]:
output11 = ""

# Write your solution here

i = 1
j = 1

while i <= 9:
    j = 1
    while j <= i:
        output11 = output11 + str(i)
        j += 1
    output11 = output11 + "\n"
    i += 1
output11 = output11.strip()
print(output11)


1
22
333
4444
55555
666666
7777777
88888888
999999999


#### Question 12&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points)
Write a Python program to display numbers that are divisible by 9 and multiple of 5, between 1000 and 2000 (including 1000 and 2000) using while loops. Create the variable `output12` and use it inside the while loop to construct the string that you want to display at the end of the program.

*Note: The displayed string should be in the same format as in Q1.*

In [57]:
output12 = ""

# Write your solution here

i = 1000

while i >= 1000 and i <= 2000:
    if i % 9 == 0 and i % 5 == 0:
        output12 = output12 + str(i) + "\n"
        i += 1
    else:
        i += 1

output12 = output12.strip()
print (output12)



1035
1080
1125
1170
1215
1260
1305
1350
1395
1440
1485
1530
1575
1620
1665
1710
1755
1800
1845
1890
1935
1980


#### Question 13&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points)
Write a Python program that iterates from 200 to 250 (including both 200 and 250) using while loops and perform the following. Create the variable `output13` and use it inside the while loop to construct the string that you want to display at the end of the program.

1. If a number is a multiply of 3, display "NUMBER is a multiply of 3!" (*Note: NUMBER should be replaced with the value at that iteration*)
2. If a number is a multiply of 5, display "NUMBER is a multiply of 5!" (*Note: NUMBER should be replaced with the value at that iteration*)
3. If a number is a multiply of 3 as well as multiply of 5, display "NUMBER is a multiply of both 3 and 5!" (*Note: NUMBER should be replaced with the value at that iteration*)

*Note: The displayed string should be in the same format as in Q1.*

In [103]:
output13 = ""

# Write your solution here

i = 200

while i <= 250:
    if i % 5 == 0 and i % 3 == 0:
        output13 = output13 + f"{i} is a multiply of both 3 and 5!" + "\n"
        i += 1
    elif i % 3 == 0:
        output13 = output13 +  f"{i} is a multiply of 3!" + "\n"
        i += 1
    elif i % 5 == 0:
        output13 = output13 +  f"{i} is a multiply of 5!" + "\n"
        i += 1
    else:
        i += 1
        
        
output13 = output13.strip()
print (output13)
    


200 is a multiply of 5!
201 is a multiply of 3!
204 is a multiply of 3!
205 is a multiply of 5!
207 is a multiply of 3!
210 is a multiply of both 3 and 5!
213 is a multiply of 3!
215 is a multiply of 5!
216 is a multiply of 3!
219 is a multiply of 3!
220 is a multiply of 5!
222 is a multiply of 3!
225 is a multiply of both 3 and 5!
228 is a multiply of 3!
230 is a multiply of 5!
231 is a multiply of 3!
234 is a multiply of 3!
235 is a multiply of 5!
237 is a multiply of 3!
240 is a multiply of both 3 and 5!
243 is a multiply of 3!
245 is a multiply of 5!
246 is a multiply of 3!
249 is a multiply of 3!
250 is a multiply of 5!


#### Question 14&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points)
Create the variable `q14_num` and assign 1912 to it. Write a Python program to reverse `q14_num` using while loop and arithmetic operations. For example, if `q14_num` is 1912, the reversed number is 2191. Create the variable `output14` and assign the reversed number to it. However, if a user has mistakenly assign one digit number (i.e., 0-9) to `q14_num`, `output14`should be the string "Please enter at least two digits to perform operations". Print `output14` at the end of the prohram to check if it works correctly.

In [10]:
q14_num = 1912

# Write your solution here
output14 = 0

if 0 <= q14_num <= 9:
    output14 =  "Please enter at least two digits to perform operations"

else:
    while q14_num > 0:
        remainder = q14_num % 10
        output14 = (output14 * 10) + remainder
        q14_num = q14_num // 10

print (output14)


2191


#### Question 15 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points) 
Write a Python program to display the series of `2, 22, 222, 2222, ..., n terms` till n terms using while loops.

For this question, create `q15_n` variable (that represents `n` in the series) and assign 20 to it. Create the variable `output15` and use it inside the while loop to construct the string that you want to display at the end of the program.

*Note: The displayed string should be in the same format as in Q1.*

In [105]:
q15_n = 20
output15 = ""

# Write your solution here
a = "2"
i = 1

while i <= q15_n:
    output15 = output15 + a * i + "\n"
    i += 1
output15 = output15.strip()
print (output15)

2
22
222
2222
22222
222222
2222222
22222222
222222222
2222222222
22222222222
222222222222
2222222222222
22222222222222
222222222222222
2222222222222222
22222222222222222
222222222222222222
2222222222222222222
22222222222222222222


#### Question 16 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points)
Write a Python program to find the sum of following series: 1 + 4 - 9 + 16 - 25 + 36 ... n terms.

For this question, create `q16_n` variable (that represents `n` in the series) and assign 20 to it. Create the variable `output16` and assign the solution to it.

In [15]:
q16_n = 20

# Write your solution here

output16 = 0
count = 0
i = 1

if (count == 0):
    output16 = output16 + (i**2)
    count += 1
    i += 1
    
while count < q16_n:       
    if count % 2 == 0:
        output16 = output16 - (i**2)
        count += 1
        i += 1
        
    else:
        output16 = output16 + (i**2)
        count += 1
        i += 1
        

print (output16)


212


#### Question 17 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)
 Write a Python program to calculate factorial using while loop. Create a variable `q17_input` and assign the number 191 to it. To calculate the factorial use the following guidelines. Create a variable `output17` and assign the solution to it.

* Factorial of number n is represented by `n!` and it is equal to `1*2*3*....*(n-1)*n`. For example, `4!` = `1*2*3*4` = 24
* Note that 1! = 1 and 0! = 1

In [22]:
q17_input = 191

# Write your solution here

output17 = 1

while q17_input > 0:
    output17 = output17 * q17_input
    q17_input -= 1

print (output17)

1848941630973752588819559184291528260234868800748408811903414244709323983074218587659329898912306172360114330860464468736986865023838872462695380757512438036901650776605459478797800350707553256022912105863632666857472153971092009065677258893759594079568220432651120989901182064402330506648926574264895332679680000000000000000000000000000000000000000000000


#### Question 18 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)
Write a Python program to display the following pattern using while loops. Create the variable `output18` and use it inside the while loop to construct the string that you want to display at the end of the program.
<p>
    &nbsp;&nbsp;&nbsp;&nbsp;*  <br>
  &nbsp;&nbsp;&nbsp;*** <br>
  &nbsp;&nbsp;***** <br>
&nbsp;******* <br>
********* <br>
&nbsp;******* <br>
&nbsp;&nbsp;***** <br>
   &nbsp;&nbsp;&nbsp;*** <br>
    &nbsp;&nbsp;&nbsp;&nbsp;*  <br>
</p>

***
<details>
  <summary><span style="color:blue">Click to see Hints</span></summary>
    
* You may need to use two while loops to print the first half and second half of the pattern.
* You may need to use `break` statement as practiced in Question 8.
</details>

In [153]:
output18 = ""

# Write your solution here
a = ""
b = ""
i = 5
j = 1
k = 5 - 1
z = " "

while i > 0:
    while j <= 9:
        a = a + k * z + "*" * j +  k * z + "\n"
        j += 2
        k -= 1
    
    i -= 1  

i = 4
k = 1
j = 7


while i > 0:
    while j >= 1:
        b = b + k * z + "*" * j +  k * z + "\n"
        j -= 2
        k += 1
        
    
    i -= 1

b = b.rstrip()
output18 = a + b

print (output18)


    *    
   ***   
  *****  
 ******* 
*********
 ******* 
  *****  
   ***   
    *


'    *    \n   ***   \n  *****  \n ******* \n*********\n ******* \n  *****  \n   ***   \n    *'

#### Question 19 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)
Write a Python program to reverse a given number using while loop. Create a variable `q19_input` and assign the number 121019200109 to it. Use the variable `output19` and assign the solution to it.

In [62]:
q19_input = 121019200109
# print (type (q19_input))

# Write your solution here


output19 = 0
if 0 <= q19_input <= 9:
    output19 = q19_input

else:   
    while q19_input > 0:    
        remainder = q19_input % 10
        output19 = (output19 * 10) + remainder
        q19_input = q19_input // 10


print (output19)
# print (type(output19))


901002910121


#### Question 20 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)
Write a Python program to find the sum of the following series using a while loop.  
$1 +\frac{y}{1!} + \frac{y^2}{2!} + ... + \frac{y^n}{n!}$

Use the following guidelines when developing the program.

* Create two variables `q20_y` and `q20_n` and assign the values 6 and 26 (Note that `q20_y` and `q20_n` represent `y` and `n` in the above-mentioned series).
* The sum should be ronded to the nearest integer value. Create the variable `output20` and assign the solution to it.

In [ ]:
q20_y, q20_n = 6, 26

# Write your solution here



#### Question 21 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)
Write a Python program to check whether a number is palindrome or not using while loop.
Palindrome is a sequence which reads the same backward as forward. Examples of palindromes are mentioned below.
* 12344321
* 2002
* 19191

Create the variable `q21_input` and assign 12123454321 to it. If `q21_input` is a palindrome assign the string "This is a palindrome" to `output21`. Otherwise, `output21` should be "This is not a palindrome"

In [26]:
q21_input = 12123454321
i = q21_input

# Write your solution here
output19 = 0

if 0 <= q21_input <= 9:
    output19 = q21_input

else:
    while q21_input > 0:
        remainder = q21_input % 10
        output19 = (output19 * 10) + remainder
        q21_input = q21_input // 10

if output19 == i:
    output21 = "This is a palindrome"
    print (output21)
    
else:
    output21 = "This is not a palindrome"
    print (output21)


This is a palindrome


#### Question 22 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)
Write a Python program that displays sum of odd numbers and even numbers separately that fall between two given numbers `q22_num1` and `q22_num2` (including both numbers) using while loop. Use the following guidelines to develop this program.

* For this question, assign the values 21 and 9 to the two variables `q22_num1` and `q22_num2`, respectively.
* Use the two variables `output22_odd` and `output22_even` to assign the sum of odd numbers and the sum of even numbers, respectively.

In [ ]:
q22_num1 = 21
q22_num2 = 9

# Write your solution here



#### Question 23 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(10 Points)
Write a Python program to display the following pattern using while loops. Create the variable `output23` and use it inside the while loop to construct the string that you want to display at the end of the program.
<p>
 &nbsp;***<br>                                                                 
 *&nbsp;&nbsp;&nbsp;*<br>                                                                  
 *&nbsp;&nbsp;&nbsp;*<br>                                                                  
 *&nbsp;&nbsp;&nbsp;*<br>                                                                  
 *&nbsp;&nbsp;&nbsp;*<br>                                                                  
 *&nbsp;&nbsp;&nbsp;*<br>                                                                  
&nbsp;*** 
</p>

In [88]:
output23 = "" 

a = "  ***  " + "\n"
b = "  ***"
i = 1

while i <= 5:
    output23 = output23 + " *"+"   "+"* "+"\n"
    i += 1

output23 = a + output23 + b
print (output23)


  ***  
 *   * 
 *   * 
 *   * 
 *   * 
 *   * 
  ***


#### Question 24 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(10 Points)
Write a Python program to display the following pattern using while loops. Create the variable `output24` and use it inside the while loop to construct the string that you want to display at the end of the program.

<p>
    * * * * *<br> 
&nbsp;* * * *<br> 
&nbsp;&nbsp;* * *<br> 
&nbsp;&nbsp;&nbsp;* *<br> 
&nbsp;&nbsp;&nbsp;&nbsp;*<br> 
&nbsp;&nbsp;&nbsp;&nbsp;*<br> 
&nbsp;&nbsp;&nbsp;* *<br> 
&nbsp;&nbsp;* * *<br> 
&nbsp;* * * *<br> 
* * * * *<br>
</p>

In [87]:
output24 = ""

# Write your solution here

i = 1
j = 1
z = " "
k = 5
while i <= 5:
    j = 1
    sum = ""
    while j <= i:
        sum = sum + k * z + "*" 
        j += 1
    print (sum)
    i += 1
    k -= 1
    

     *
    *    *
   *   *   *
  *  *  *  *
 * * * * *
